File: tests/e2e/assign_2d.c
#include <math.h>
#include <stdio.h>

extern void assign_2d(float *dst, float *src);

int main() {
    float src[16], dst[16] = {0};
    for (int i = 0; i < 16; i++)
        src[i] = (float)i;
    assign_2d(dst, src);
    int ok = 1;
    for (int i = 0; i < 16; i++) {
        if (fabsf(dst[i] - src[i]) > 1e-5f) {
            ok = 0;
            break;
        }
    }
    printf("%s\n", ok ? "OK" : "FAIL");
    return 0;
}

File: tests/e2e/assign_2d.py
# RUN: uv run xdsl-exo -o %t.mlir %s
# RUN: mlir-opt --convert-arith-to-llvm --convert-cf-to-llvm --convert-func-to-llvm --reconcile-unrealized-casts %t.mlir | mlir-translate --mlir-to-llvmir | llc -filetype=obj -o %t.o
# RUN: clang %t.o %S/assign_2d.c -o %t.bin -lm
# RUN: %t.bin | filecheck %s

from __future__ import annotations

from exo import *


# CHECK: OK
@proc
def assign_2d(dst: f32[4, 4] @ DRAM, src: f32[4, 4] @ DRAM):
    for i in seq(0, 4):
        for j in seq(0, 4):
            dst[i, j] = src[i, j]
